curp/src/client_new/unary.rs

```diff
@@ -0,0 +1,620 @@
+                    // We need to retry as same as a server error.
+                    err = Some(CurpError::from(ser_err));
+                    continue;
+                }
+            };
+            if let Err(cmd_err) = dr {
+                // got a command execution error early, abort the next requests and return the cmd error
+                return Ok(Err(cmd_err));
+            }
+            // if the propose meets the super quorum and we got the execute result,
+            // that means we can safely abort the next requests
+            if ok_cnt >= super_quorum {
+                if let Some(er) = execute_result {
+                    debug!("fast round for cmd({}) succeed", propose_id);
+                    return Ok(Ok(er));
+                }
+            }
+        }
+
+        if let Some(err) = err {
+            return Err(err);
+        }
+
+        // We will at least send the request to the leader if no `WrongClusterVersion` returned.
+        // If no errors occur, the leader should return the ER
+        // If it is because the super quorum has not been reached, an error will definitely occur.
+        unreachable!("leader should return ER if no error happens");
+    }
+
+    /// Wait synced result from server
+    async fn slow_round(
+        &self,
+        propose_id: ProposeId,
+    ) -> Result<Result<(C::ASR, C::ER), C::Error>, CurpError> {
+        let timeout = self.config.wait_synced_timeout;
+        let req = WaitSyncedRequest::new(propose_id, self.cluster_version());
+        let resp = self
+            .for_leader(|conn| async move { conn.wait_synced(req, timeout).await })
+            .await??
+            .into_inner();
+        let synced_res = resp.map_result::<C, _, _>(|res| res).map_err(|ser_err| {
+            warn!("serialize error: {ser_err}");
+            // Same as fast round, we blame the server for the serializing error.
+            CurpError::from(ser_err)
+        })?;
+        Ok(synced_res)
+    }
+
+    /// Get the client id
+    #[allow(clippy::unused_async)] // TODO: grant a client id from server
+    async fn get_client_id(&self) -> Result<u64, CurpError> {
+        Ok(rand::random())
+    }
+
+    /// New a seq num and record it
+    #[allow(clippy::unused_self)] // TODO: implement request tracker
+    fn new_seq_num(&self) -> u64 {
+        rand::random()
+    }
+
+    /// Generate a new propose id
+    async fn gen_propose_id(&self) -> Result<ProposeId, CurpError> {
+        let client_id = self.get_client_id().await?;
+        let seq_num = self.new_seq_num();
+        Ok(ProposeId(client_id, seq_num))
+    }
+}
+
+#[async_trait]
+impl<C: Command> ClientApi for Unary<C> {
+    /// The error is generated from server
+    type Error = CurpError;
+
+    /// The command type
+    type Cmd = C;
+
+    /// Get the local connection when the client is on the server node.
+    fn local_connect(&self) -> Option<Arc<dyn ConnectApi>> {
```

Comment:
Do we still needs this `local_connect` if we have the bypassed connect?

Reply:
This method find the bypassed connect (if exist) in `connects`. If found, we could use it without taking care of network issues if user enable serializable options.

Reply:
Currently, it is used within the fetch cluster. It will attempt to fetch from the local server first, and then fetch from the cluster.